File: middleware-moshi/src/main/java/com/nytimes/android/external/store/middleware/moshi/MoshiStringParser.java
package com.nytimes.android.external.store.middleware.moshi;

import com.nytimes.android.external.store.base.Parser;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

import java.io.IOException;
import java.lang.reflect.Type;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import io.reactivex.annotations.NonNull;

public class MoshiStringParser<Parsed> implements Parser<String, Parsed> {

    private final JsonAdapter<Parsed> jsonAdapter;

    @Inject
    public MoshiStringParser(@Nonnull Moshi moshi, @Nonnull Type type) {
        jsonAdapter = moshi.adapter(type);
    }


    @Override
    @SuppressWarnings({"PMD.EmptyCatchBlock"})
    public Parsed apply(@NonNull String s) {
        try {
            return jsonAdapter.fromJson(s);
        } catch (IOException e) {
            return null;
        }
    }
}

File: middleware-moshi/src/main/java/com/nytimes/android/external/store3/middleware/moshi/MoshiStringParser.kt
package com.nytimes.android.external.store3.middleware.moshi

import com.nytimes.android.external.store3.base.Parser
import com.nytimes.android.external.store3.util.ParserException
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi

import java.io.IOException
import java.lang.reflect.Type
import javax.inject.Inject

import io.reactivex.annotations.NonNull

class MoshiStringParser<Parsed> @Inject
constructor(moshi: Moshi, type: Type) : Parser<String, Parsed> {

    private val jsonAdapter: JsonAdapter<Parsed>

    init {
        jsonAdapter = moshi.adapter(type)
    }

    @Throws(ParserException::class)
    override suspend fun apply(@NonNull s: String): Parsed {
        try {
            return jsonAdapter.fromJson(s)!!
        } catch (e: IOException) {
            throw ParserException(e.message, e)
        }

    }
}
